packages/protocol/src/utils/SPID.ts

```diff
@@ -0,0 +1,202 @@
+
+class SPIDValidationError extends Error {
+    data: SPIDLikePartial
+    constructor(msg: string, data: SPIDLikePartial) {
+        super(format(msg, data))
+        this.data = data
+        if (Error.captureStackTrace) {
+            Error.captureStackTrace(this, this.constructor)
+        }
+    }
+}
+
+/**
+ * SPID - Stream Partition ID
+ */
+export default class SPID {
+    /** stream id */
+    public readonly id: string
+    /** stream partition */
+    public readonly partition: number
+    /** toString/fromString separator */
+    protected static readonly SEPARATOR = '#'
+
+    /** string key representing SPID */
+    public readonly key: string
+
+    /**
+     * @param id - stream id
+     * @param partition - stream partition
+     */
+    constructor(id: string, partition: number) {
+        this.id = id
+        this.partition = partition
+        this.validate()
+
+        // static cached values to prevent unnecessary runtime allocations
+        this.key = `${this.id}${SPID.SEPARATOR}${this.partition}`
+        // prevent all mutation
+        Object.freeze(this)
+    }
+
+    /**
+     * Throws if data is invalid.
+     */
+    private validate() {
+        if (typeof this.id !== 'string' || this.id.length === 0) {
+            throw new SPIDValidationError('SPID validation failed: id must be non-empty string: %o', {
+                id: this.id,
+                partition: this.partition
+            })
+        }
+
+        if (typeof this.partition !== 'number' || !Number.isSafeInteger(this.partition) || this.partition < 0) {
+            throw new SPIDValidationError('SPID validation failed: partition must be a safe integer >= 0: %o', {
+                id: this.id,
+                partition: this.partition
+            })
+        }
+    }
+
+    /**
+     * True iff other value is equivalent.
+     */
+    equals(other: SPIDLike): boolean {
+        // check if same instance
+        if (other === this) { return true }
+        let otherSpid: SPID
+        try {
+            otherSpid = SPID.from(other)
+        } catch (_err) {
+            // ignore error
+            // not equal if not valid
+            return false
+        }
+
+        // check key matches
+        return this.key === otherSpid.key
+    }
+
+    /**
+     * Convert SPIDLikePartial to SPIDObjectPartial
+     * i.e. normalizes various input types/shapes to { streamId?, streamPartition? }
+     * Note: does not throw on malformed input
+     */
+    static toSPIDObjectPartial(spidLike: SPIDLikePartial): SPIDObjectPartial {
+        // convert from string
+        if (typeof spidLike === 'string') {
+            const [streamId, partitionStr] = spidLike.split(this.SEPARATOR)
+            // partition is optional, falls back to undefined i.e. default
+            const streamPartition = partitionStr != null ? Number.parseFloat(partitionStr) : undefined
```

Comment:
Could use `parseInt` instead of `parseFloat`? (also at line 125)

Reply:
If user tries to use partition 1.5 should we error or pick partition 1?

Reply:
My thinking:
If it were a decimal given by the user:
- It's most likely a typing / copy-paste errors of sort => nice to provide an error to the user so they can have a look at it and make sure they are getting what they intended.
- If they are genuinely giving a decimal they have misunderstood something. Then also an error maybe useful to course correct their understanding.